server/src/db/lifecycle/drop.rs

```diff
@@ -80,3 +80,110 @@ pub fn drop_chunk(
 
     Ok((tracker, fut.track(registration)))
 }
+
+pub fn drop_partition(
+    partition: LifecycleWriteGuard<'_, Partition, LockableCatalogPartition>,
+) -> Result<(
+    TaskTracker<Job>,
+    TrackedFuture<impl Future<Output = Result<()>> + Send>,
+)> {
+    let db = Arc::clone(&partition.data().db);
+    let preserved_catalog = Arc::clone(&db.preserved_catalog);
+    let table_name = partition.table_name().to_string();
+    let partition_key = partition.key().to_string();
+
+    let (tracker, registration) = db.jobs.register(Job::DropPartition {
+        partition: partition.addr().clone(),
+    });
+
+    // get locks for all chunks
+    let lockable_chunks: Vec<_> = partition
+        .chunks()
+        .map(|chunk| LockableCatalogChunk {
+            db: Arc::clone(&db),
+            chunk: Arc::clone(chunk),
+        })
+        .collect();
+    let mut guards: Vec<_> = lockable_chunks
+        .iter()
+        .map(|lockable_chunk| lockable_chunk.write())
+        .collect();
+
+    // NOTE: here we could just use `drop_chunk` for every chunk, but that would lead to a large number of catalog
+    // transactions and is rather inefficient.
+
+    // pre-check all chunks before touching any
+    for guard in &guards {
+        // check if we're dropping an unpersisted chunk in a persisted DB
+        // See https://github.com/influxdata/influxdb_iox/issues/2291
+        if db.rules().lifecycle_rules.persist
+            && !matches!(guard.stage(), ChunkStage::Persisted { .. })
+        {
+            return CannotDropUnpersistedChunk {
+                addr: guard.addr().clone(),
+            }
+            .fail();
+        }
+
+        if let Some(action) = guard.lifecycle_action() {
+            return Err(Error::ChunkError {
+                source: crate::db::catalog::chunk::Error::LifecycleActionAlreadyInProgress {
+                    chunk: guard.addr().clone(),
+                    lifecycle_action: action.metadata().name().to_string(),
+                },
+            });
+        }
+    }
+
+    // start lifecycle action
+    for guard in &mut guards {
+        guard.set_dropping(&registration)?;
+    }
+
+    // Drop locks
+    let chunks: Vec<_> = guards
+        .into_iter()
+        .map(|guard| guard.into_data().chunk)
+        .collect();
+    let partition = partition.into_data().partition;
+
+    let fut = async move {
+        debug!(%table_name, %partition_key, "dropping partition");
+
+        // collect parquet files that we need to remove
+        let mut paths = vec![];
+        for chunk in &chunks {
+            let chunk_read = chunk.read();
+
+            if let ChunkStage::Persisted { parquet, .. } = chunk_read.stage() {
+                let path: DirsAndFileName = parquet.path().into();
+                paths.push(path);
+            }
```

Comment:
I would suggest something like

```suggestion
            }
            else if db.rules().lifecycle_rules.persist {
              unreachable!()
            }
```

to both:
1. Make it clear to future readers that the assumption of stage is valid as it was checked above
2. If the check above is ever changed, will assert if they don't also update this code (which relies on the assumption)

Reply:
Done (same for `drop_chunk`).